cocos/particle/particle-culler.ts

```diff
@@ -0,0 +1,407 @@
+
+        if (ps.simulationSpace === Space.Local) {
+            const r:Quat = ps.node.getRotation();
+            Mat4.fromQuat(this._localMat, r);
+            this._localMat.transpose(); // just consider rotation, use transpose as invert
+        }
+
+        for (let i = 0; i < particleLst.length; ++i) {
+            const p: Particle = particleLst[i];
+            p.remainingLifetime -= dt;
+            Vec3.set(p.animatedVelocity, 0, 0, 0);
+
+            if (ps.simulationSpace === Space.Local) {
+                const gravityFactor = -ps.gravityModifier.evaluate(1 - p.remainingLifetime / p.startLifetime, pseudoRandom(p.randomSeed))! * 9.8 * dt;
+                this._gravity.x = 0.0;
+                this._gravity.y = gravityFactor;
+                this._gravity.z = 0.0;
+                this._gravity.w = 1.0;
+                this._gravity = this._gravity.transformMat4(this._localMat);
+
+                p.velocity.x += this._gravity.x;
+                p.velocity.y += this._gravity.y;
+                p.velocity.z += this._gravity.z;
+            } else {
+                // apply gravity.
+                p.velocity.y -= ps.gravityModifier.evaluate(1 - p.remainingLifetime / p.startLifetime, pseudoRandom(p.randomSeed))! * 9.8 * dt;
+            }
+
+            Vec3.copy(p.ultimateVelocity, p.velocity);
+
+            this._runAnimateList.forEach((value) => {
+                value.animate(p, dt);
+            });
+
+            Vec3.scaleAndAdd(p.position, p.position, p.ultimateVelocity, dt); // apply velocity.
+        }
+    }
+
+    private _calculateBounding (isInit: boolean) {
+        const size: Vec3 = new Vec3();
+        const position: Vec3 = new Vec3();
+        const subPos: Vec3 = new Vec3();
+        const addPos: Vec3 = new Vec3();
+
+        const meshSize: Vec3 = new Vec3(1.0, 1.0, 1.0);
+        if (this._processor.getInfo()!.renderMode === RenderMode.Mesh) {
+            const mesh: Mesh | null = this._processor.getInfo().mesh;
+            if (mesh && mesh.struct.minPosition && mesh.struct.maxPosition) {
+                const meshAABB: AABB = new AABB();
+                AABB.fromPoints(meshAABB, mesh.struct.minPosition, mesh.struct.maxPosition);
+                const meshMax = Math.max(meshAABB.halfExtents.x, meshAABB.halfExtents.y, meshAABB.halfExtents.z);
+                meshSize.set(meshMax, meshMax, meshMax);
+            }
+        }
+
+        for (let i = 0; i < this._particlesAll.length; ++i) {
+            const p: Particle = this._particlesAll[i];
+            Vec3.multiply(size, _node_scale, p.size);
+            Vec3.multiply(size, size, meshSize);
+            position.set(p.position);
+            if (this._particleSystem.simulationSpace !== Space.World) {
+                Vec3.transformMat4(position, position, this._particleSystem.node._mat);
+            }
+            if (isInit && i === 0) {
+                this.minPos.set(position.x - size.x, position.y - size.y, position.z - size.z);
```

Comment:
Vec3.subtract(this.minPos, position, size);
Vec3.add(this.maxPos, position, size);